### ISSUE TITLE
Need to retain a predictable query result ordering even in the absence of ORDER BY.

### ISSUE THREAD

**Issue by jeenbroekstra** [eclipse/rdf4j#487]
(Migrated from https://openrdf.atlassian.net/browse/SES-1101)

From http://www.w3.org/TR/2012/WD-sparql11-query-20120724/#modOffset : "Using LIMIT and OFFSET to select different subsets of the query solutions will not be useful unless the order is made predictable by using ORDER BY." In practice, the standard Sesame backends always return results in the same order. This is not documented behavior, however, nor is it guaranteed to remain this way in future releases. If changes ever are made to the query engine that throw away the predictable ordering, it is suggested here that it would be a good idea to add something to the API to request some sort of canonical ordering in the absence of ORDER BY, otherwise paging through query results in Workbench would be broken. (See linked issue SES-376..)

**Comment by barthanssens** [comment]
Also important for implementing Linked Data Fragments

**Comment by jeenbroekstra** [comment]
In my view, this is not something we can support. RDF4J is an API for multiple database implementations. Enforcing a canonical ordering at the level of the RDF4J APIs would require significant changes not just in our own database implementations, but in _every_ third party triplestore exposed through RDF4J.

Furthermore there are solid reasons for not enforcing this, both conceptual, and practical. Most importantly is that forcing order is a significant performance burden, and not something we want to do for _every_ SPARQL query. Also: forcing ordering means that database implementations have less freedom to optimize the query plan, switch to the most optimal index, etc. etc. 

In short: my view is that this is neither realistic nor desirable.

**Comment by jeenbroekstra** [comment]
A possible compromise, perhaps, is that we introduce a default ordering only when limit/offset are used.
